### ISSUE TITLE
Support cleanup between tests with multiple databases

### ISSUE THREAD

**Issue by ftobia** [pytest-dev/pytest-django#76]
pytest-django doesn't clean up between tests when using Django with multiple databases. The problem is related to this StackOverflow question: http://stackoverflow.com/questions/10121485/django-testcase-not-using-transactions-on-secondary-database

The `db` fixture uses Django's TestCase under the covers.

Here is the code from the original `db` fixture:

```
case = TestCase(methodName='__init__')
case._pre_setup()
request.addfinalizer(case._post_teardown)
request.addfinalizer(_django_cursor_wrapper.disable)
```

Here is how I patched it as a work-around in my code:

```
case = TestCase(methodName='__init__')
case.multi_db = True
case._pre_setup()
request.addfinalizer(case._post_teardown)
request.addfinalizer(_django_cursor_wrapper.disable)
```

Obviously that multi_db flag exists and defaults to False for a reason. I'm not sure the right way to incorporate multi_db support into pytest_django, and I'm not sure the right way to test such a change.

If you have a suggestion I can work on a pull request.

**Comment by pelme** [comment]
@slafs There has not been much discussions/requests for multi database support apart from this.

It would be pretty simple to pass on `multi_db=True`, but we also needs some kind of tests for this, so we also need a secondary database set up. If anyone is willing to work on this, it would be great :)

**Comment by robdennis** [comment]
I've been using the work-around @ftobia mentions in the opening post, but after upgrading to Django 1.8.4 from 1.7 it stopped working.

There was Django [change](https://github.com/django/django/commit/da9fe5c717c179a9e881a40efd3bfe36a21bf4a6) to make the code handling setup/tearDown look for a ``multi_db`` class attribute instead of an instance attribute.

Changing @ftobia's work-around to the following worked for me:
```
case = TestCase(methodName='__init__')
case.__class__.multi_db = True
case._pre_setup()
request.addfinalizer(case._post_teardown)
request.addfinalizer(_django_cursor_wrapper.disable)
```

**Comment by adrianandreias** [comment]
+1 :+1: 

This is stopping us from switching from Django test runner to pytest.

**Comment by eremeevdev** [comment]
Why no use `django_case.multi_db = True` here https://github.com/pytest-dev/pytest-django/blob/master/pytest_django/fixtures.py#L107 by default?

**Comment by knabben** [comment]
+1

**Comment by guilhermemaba** [comment]
+1

**Comment by pelme** [comment]
I think that having a marker on the django_db marker like would by I good API:

```python
@pytest.mark.django_db(transactional=True, multi_db=True)
def test_a():
    pass
```

The implementation should be relatively similar to `reset_sequences` (PR #308) or `serialized_rollback` (PR #353).

The "tricky" part would be to improve pytest-django's internal test suite to contain multiple databases.

If someone wants to work in this I will for sure review it and help in getting it merged. Feel free to start working on a PR if this is something that would be interesting to you!

**Comment by blueyed** [comment]
Left some comments at https://github.com/pytest-dev/pytest-django/pull/397#issuecomment-261987751, and created a PR, which would allow to change this in a generic way: https://github.com/pytest-dev/pytest-django/pull/431.

**Comment by cheniel** [comment]
Another option to enable multi_db for those looking for a temporary solution. Put this in conftest.py
```
def pytest_sessionstart(session):
    from django.test import TransactionTestCase
    TransactionTestCase.multi_db = True
```

**Comment by jcushman** [comment]
For anyone else looking for workarounds, note that `multi_db` was deprecated in Django 2.2. The replacement that worked for me is:

`TransactionTestCase.databases = set(settings.DATABASES.keys())`

**Comment by merit-mthompson** [comment]
@jcushman do you mind sharing the full fixture or code snippet you used?

**Comment by merit-mthompson** [comment]
For whatever reason, `pytest_sessionstart()` in our base `conftest.py` wasn't doing anything. Took a page from #342 along with standard pytest `monkeypatch` stuff and have landed on the following for our multi-database use:

```python
@pytest.fixture(autouse=True)
def django_db_multiple(monkeypatch, settings):
    """
    Ensure all test functions using Django test cases have multiple database
    support. This is mostly/only so that Django will wrap ALL database use with
    atomic blocks like it does for DEFAUL_DB_ALIAS.

    https://github.com/django/django/blob/master/django/test/testcases.py#L903
    """
    from django.test import TestCase
    from django.test import TransactionTestCase

    db_keys = set(settings.DATABASES.keys())

    monkeypatch.setattr(TestCase, 'databases', db_keys)
    monkeypatch.setattr(TransactionTestCase, 'databases', db_keys)
```

**Comment by theoden-dd** [comment]
In order not to loose pytest superpowers by switching to unittest Django test cases, I copied and patched pytest-django internals to this ugly (yet working!) hack:

```python
from typing import Optional, Type, TypeVar

import pytest
from django.test import TransactionTestCase
from pytest_django.django_compat import is_django_unittest

TestCase = TypeVar('TestCase', bound=TransactionTestCase)


def _django_db_fixture_helper(
    request, django_db_blocker, transactional: bool = False, reset_sequences: bool = False,
) -> Optional[Type[TestCase]]:
    if is_django_unittest(request):
        return None

    if not transactional and 'live_server' in request.fixturenames:
        # Do nothing, we get called with transactional=True, too.
        return None

    django_db_blocker.unblock()
    request.addfinalizer(django_db_blocker.restore)

    if transactional:
        from django.test import TransactionTestCase as DjangoTestCase  # noqa: WPS433

        if reset_sequences:

            class ResetSequenceTestCase(DjangoTestCase):  # noqa: WPS431
                reset_sequences = True

            DjangoTestCase = ResetSequenceTestCase  # type: ignore[misc] # noqa: N806
    else:
        from django.test import TestCase as DjangoTestCase  # type: ignore[no-redef] # noqa: WPS433

    return DjangoTestCase  # type: ignore[return-value]


@pytest.fixture()
def db_case(request, django_db_setup, django_db_blocker):
    """Require a django test database.

    This database will be setup with the default fixtures and will have
    the transaction management disabled. At the end of the test the outer
    transaction that wraps the test itself will be rolled back to undo any
    changes to the database (in case the backend supports transactions).
    This is more limited than the ``transactional_db`` resource but
    faster.

    If multiple database fixtures are requested, they take precedence
    over each other in the following order (the last one wins): ``db``,
    ``transactional_db``, ``django_db_reset_sequences``.
    """
    if 'django_db_reset_sequences' in request.fixturenames:
        request.getfixturevalue('django_db_reset_sequences')
    if (
        'transactional_db' in request.fixturenames
        or 'live_server' in request.fixturenames
    ):
        request.getfixturevalue('transactional_db')
    else:
        django_case: Optional[Type[TransactionTestCase]] = _django_db_fixture_helper(
            request, django_db_blocker, transactional=False,
        )

        def factory(dbs=None):  # noqa: WPS430
            if django_case is None:
                return
            CaseType: Type[TransactionTestCase] = django_case  # noqa: N806
            if dbs is not None:
                class DatabasesSetTestCase(  # noqa: WPS431
                    CaseType,  # type: ignore[valid-type, misc]
                ):
                    databases = dbs
                CaseType = DatabasesSetTestCase  # noqa: N806
            test_case: TransactionTestCase = CaseType(methodName='__init__')
            test_case._pre_setup()  # type: ignore[attr-defined] # noqa: WPS437
            request.addfinalizer(
                test_case._post_teardown,  # type: ignore[attr-defined] # noqa: WPS437
            )

        return factory
```

With it one can use something like:
```python
class TestCase(object):

    def test_ok(
        self,
        db_case,
        ...
    ):
        db_case(dbs=('non_default_db_alias',))
```

**Comment by roks0n** [comment]
> @jcushman do you mind sharing the full fixture or code snippet you used?

Adding reply in case anyone else will bump into this.

You need to add the code below into the `conftest.py`. This fixture gets picked up automatically so no need to add it to a test.

```py
def pytest_sessionstart(session):
    from django.test import TransactionTestCase

    TransactionTestCase.databases = set(settings.DATABASES.keys())
````

**Comment by bluetech** [comment]
So I believe this issue comes down to proper multi-db support. I'll close this issue as duplicate of that.
